docs: tw700-sequencer-config-ref (#3529)
OffchainLabs/arbitrum-docs: docs/launch-arbitrum-chain/operate/sequencer-troubleshooting.mdx

</VanillaAdmonition>

## Feed relay architecture

The <a data-quicklook-from="sequencer-feed">sequencer feed</a> is a WebSocket stream of transactions in the order the <a data-quicklook-from="sequencer">sequencer</a> chose. It gives you data before batches land on the <a data-quicklook-from="parent-chain">parent chain</a>, which is why nodes use it to stay current instead of waiting for batches.

### What produces the feed

Any <a data-quicklook-from="arbitrum-nitro">Nitro</a> node with feed output enabled runs a broadcast server on port `9642` and serves the `/feed` endpoint. Two roles enable it:

| Role      | What it does                                                 | Configuration                          |
| --------- | ------------------------------------------------------------ | -------------------------------------- |
| Sequencer | Broadcasts messages it sequences itself                      | `node.feed.output.enable=true`         |
| Relay     | Subscribes to an upstream feed and rebroadcasts it unchanged | `node.feed.input.url` plus feed output |

> A relay does not sequence, validate, or execute anything. It reads an upstream feed and distributes it to more subscribers. The message content is identical either way, so a client cannot tell from the payload whether it is connected to a sequencer or a relay.

Reviewer: ```suggestion
A relay does not sequence, validate, or execute anything. It reads an upstream feed and distributes it to more subscribers. The message content is identical either way, so a <a data-quicklook-from="client">client</a> cannot tell from the payload whether it is connected to a sequencer or a relay.
```

### Sequencer feed compared to relay feed

The difference is operational, not structural:

- **Load isolation.** Every subscriber on a sequencer is a connection the sequencer maintains instead of ordering transactions. Relays absorb that load.

> - **Compression.** Public feed endpoints may serve compressed messages using a custom dictionary. A local relay serves an uncompressed feed by default, which is why you should run one if you consume the feed with anything other than a standard Nitro node.

Reviewer: ```suggestion
- **Compression.** Public feed endpoints may serve compressed messages using a custom dictionary. A local relay serves an uncompressed feed by default, so if you consume the feed with anything other than a standard Nitro node, run one.
```

| `node.feed.output.ping`                     | Server | 5s      | How often the server pings each client                        |
| `node.feed.output.client-timeout`           | Server | 15s     | How long the server waits before dropping a silent client     |
| `node.feed.output.read-timeout`             | Server | 1s      | How long the server waits to read data (including pings)      |
| `node.feed.input.timeout`                   | Client | 20s     | How long the client waits for data before treating it as dead |
| `node.feed.input.reconnect-initial-backoff` | Client | 1s      | First reconnect delay                                         |
| `node.feed.input.reconnect-maximum-backoff` | Client | 64s     | Reconnect delay ceiling; the delay doubles each attempt       |

## Broadcast backlog errors

The broadcast server keeps recent messages in a **backlog** so a reconnecting client can catch up without a full resync. The backlog is split into segments holding `node.feed.output.backlog.segment-limit` messages each (default `240`).

### `error in backlogSegment type assertion: clearing backlog`

> This is logged at `ERROR`. Despite the wording, it is a broadcaster-side memory structure problem, not chain data loss or corruption. Blocks the sequencer already produced are unaffected.

Reviewer: ```suggestion
Nitro logs this at `ERROR`. Despite the wording, it is a broadcaster-side memory structure problem, not chain data loss or corruption. Blocks the sequencer already produced are unaffected.
```

**What happens.** The backlog empties itself. Connected clients can no longer catch up from memory, so they resync from their configured source. Feed consumers see a gap and reconnect.

**Where it comes from.** The message appears at five places in `broadcaster/backlog/backlog.go`, and they do not all behave the same way:

- **Four sites inside `delete(confirmed)`** call `reset()` and genuinely clear the backlog. This runs when the broadcaster prunes messages it has seen confirmed on the parent chain.
- **One site inside `IsBacklogSegmentNil`** logs the identical text but clears nothing—it only reports a failed type check and returns. If you see this message and the backlog is still populated, you hit this path.

Because the text is the same in both cases, do not infer a backlog reset from the log line alone. Confirm it with `arb/feed/backlog/messages`, which drops to zero on a real reset.

One related message comes from the same code path.

#### `confirmed sequence number is past the end of stored messages`

- [WARN]—The confirmed message number is ahead of everything the backlog holds, so it resets.

> - [CAUSE]—The broadcaster fell far behind, or restarted and is being told about confirmations for messages it never buffered.

Reviewer: ```suggestion
- [CAUSE]—The broadcaster fell far behind, or it restarted and received confirmations for messages it never buffered.
```

</VanillaAdmonition>

## Block production halt conditions

Block production stops for one of four reasons: no sequencer holds the lockout, the sequencer refuses to build a block, the sequencer accepts no transactions, or execution fails. The checks below run in that order inside Nitro's block creation loop.

### 1. No sequencer is chosen

With the coordinator enabled, exactly one sequencer holds the lockout in Redis. When none does, every node forwards transactions to a target that is not sequencing, and the chain stops.

#### `sequencer priorities unset`

- [ERROR]—The `coordinator.priorities` key does not exist in Redis.
- [CAUSE]—A new or wiped Redis instance, or one that was never populated.
- [ACTION]—Register your sequencers. See [Redis priority registration](/launch-arbitrum-chain/run-a-node/high-availability-sequencer.mdx#redis-priority-registration).

#### `no sequencer appears to want the lockout on redis`

> - [DEBUG→WARN→ERROR]—The priority list exists, but no sequencer on it has set its wants-lockout key. Logged at `DEBUG` at first, `WARN` after 10 seconds, `ERROR` after 20, and throttled to once every 5 seconds.

Reviewer: ```suggestion
- [DEBUG→WARN→ERROR]—The priority list exists, but no sequencer on it has set its wants-lockout key. Nitro logs this at `DEBUG` at first, escalates to `WARN` after 10 seconds and `ERROR` after 20, and throttles it to once every 5 seconds.
```

- [CAUSE]—Every listed sequencer is unsynced, down, or unreachable—or the URLs in the list do not match any running sequencer's `my-url`.
- [ACTION]—Compare the `priorities` value printed in the log against each sequencer's configured `my-url`. They must match exactly.

#### `sequencer is not synced`

> - [WARN]—The sequencer will not ask for the lockout because it has not caught up. Sync details are attached to the log entry.

Reviewer: ```suggestion
- [WARN]—The sequencer does not ask for the lockout because it has not caught up. Nitro attaches sync details to the log entry.
```

- [CAUSE]—The node is still replaying messages, or it cannot reach the feed or parent chain.
- [MEANING]—Expected during startup. Sustained means the node cannot catch up.

#### `myurl main sequencer, but no sequencer exists`

- [ERROR]—The node is top of the priority list, but it has no sequencer configured.

> - [CAUSE]—A node registered in `coordinator.priorities` without `node.sequencer.enable=true`. Registering a batch poster causes exactly this.

Reviewer 2: ```suggestion
- [CAUSE]—A node registered in `coordinator.priorities` without `node.sequencer=true`. Registering a batch poster causes exactly this.
```

- [MEANING]—Two very different faults share this message: a parent chain connection that is down or lagging, and a local clock that has drifted.
- [ACTION]—Check parent chain RPC reachability and freshness first, then verify NTP on the sequencer host. Raising the delta hides the symptom without fixing either cause.

<VanillaAdmonition type="caution" title="This check only runs with a parent chain reader">

The timestamp check applies only when the sequencer has a parent chain reader configured. A sequencer without one never halts for this reason, which is why local dev nodes do not reproduce it.

</VanillaAdmonition>

#### `sequencer block creation panicked`

- [ERROR]—Block creation panicked. Nitro recovers, returns an internal error to every queued transaction, and waits `max-block-speed` before retrying.
- [MEANING]—The sequencer survives, so this appears as a stall rather than a crash. A repeating panic halts the chain in practice.
- [ACTION]—Capture the attached backtrace and report it.

### 3. The sequencer accepts no transactions

> The sequencer keeps running here, but the block it builds is empty because everything is rejected before it is queued.

Reviewer: ```suggestion
The sequencer keeps running here, but the block it builds is empty because the sequencer rejects everything before queuing it.
```

| Rejection message                                                                    | Trigger                                                                       |
| ------------------------------------------------------------------------------------ | ----------------------------------------------------------------------------- |
| `currently not accepting transactions due to expected surplus being below threshold` | Expected surplus fell below `expected-surplus-hard-threshold`                 |
| `transaction sender is not on the whitelist`                                         | `sender-whitelist` is set and the sender is absent                            |

> | `sequencer temporarily not available`                                                | This node is forwarding rather than sequencing, and its forwarder is disabled |

Reviewer: ```suggestion
| `sequencer temporarily not available`                                                | This node is forwarding rather than sequencing, and its <a data-quicklook-from="forwarder">forwarder</a> is disabled |
```
